### PR TITLE
Migrate to the latest version of the events

### PR DESCRIPTION
**Pull request by tommpy**

### CODE REVIEW

File: src/main/scala/com/ovoenergy/event/gateway/manager/event/package.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  type LinkClickedLatest = LinkClickedV2
+  def LinkClickedLatest(metadata: MetadataV2,
+                        internalMetadata: InternalMetadata,
+                        gateway: Gateway,
+                        url: String): LinkClickedLatest = {
+    LinkClickedV2(metadata, internalMetadata, gateway, url)
+  }
+
+  type EmailProgressedLatest = EmailProgressedV2
+  def EmailProgressedLatest(metadata: MetadataV2,
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/event/gateway/manager/event/package.scala]: I'm not so sure the type alias really buys us anything, given that you had to reimplement `EmailProgressedV2.apply` here.

File: src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/DynamoFormats.scala [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    DynamoFormat.iso[Instant, Long](x => Instant.ofEpochMilli(x))(_.toEpochMilli)
+
+  import io.circe._
+  import io.circe.parser._
+  import io.circe.generic.semiauto._
+  import io.circe.generic.auto._
+  import io.circe.shapes._
+
+  private implicit val deliverToCirceDecoder: Decoder[DeliverTo] = deriveDecoder[DeliverTo]
+  private implicit val deliverToCirceEncoder: Encoder[DeliverTo] = deriveEncoder[DeliverTo]
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/DynamoFormats.scala]: This Scanamo-via-circe stuff makes me sad. I'll experiment with Scanamo now and try to find out why we need this workaround.

File: src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/DynamoFormats.scala [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        case Left(error) =>
+          logWarn(Map(), "Unable to deserialize to DeliverTo", error)
+          throw new IllegalArgumentException("Unable to deserialize to DeliverTo", error)
+      }
+    }
+  )(
+    (deliverTo) => deliverToCirceEncoder.apply(deliverTo).spaces2
+  )
+
+  override def loggerName: String = "com.ovoenergy.event.gateway.manager.dynamo.DynamoFormats"
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/DynamoFormats.scala]: Not a blocker for this PR, but we should automate this like we have done in other projects

File: src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/Persistence.scala [hunk 4]
```diff
@@ -48,7 +48,10 @@ object Persistence {
 class Persistence(context: Context) extends LoggingWithMDC {
 
   import Persistence._
+  import DynamoFormats._
+  //implicit lazy val format = DynamoFormat.xmap[GatewayIdentifier]
 
+  val f = implicitly[DynamoFormat[GatewayIdentifier]]
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/Persistence.scala]: Is this line needed? `f` doesn't seem to be used

File: src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/package.scala [hunk 5]
```diff
@@ -31,7 +33,7 @@ package object dynamo {
 
   case class GatewayIdentifier(
       id: GatewayIdentifier.Id,
-      metadata: Metadata,
+      metadata: MetadataV2,
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/package.scala]: I think we might have a problem here. Anything stored to Dynamo before this change will include a `Metadata`, but after this change we will try to read it back from the DB as a `MetadataV2`, which will fail.

Not sure of the best solution to this. Include both v1 and v2, and make them optional? e.g.

```scala
case class GatewayIdentifier(
  id: GatewayIdentifier.Id,
  metadata: Option[Metadata],
  metadataV2: Option[MetadataV2],
  internalMetadata: InternalMetadata
)
```

That's pretty horrible.

File: src/main/scala/com/ovoenergy/event/gateway/manager/kafka/IssuedForDeliveryGraph.scala [hunk 6]
```diff
@@ -20,13 +18,15 @@ import scala.util.control.NonFatal
...
       config: Config,
       topic: String,
-      persistEvent: (IssuedForDelivery) => Unit,
-      sendCommProgressedEvent: (IssuedForDelivery => Future[_])
-  )(implicit actorSystem: ActorSystem, ec: ExecutionContext): RunnableGraph[Consumer.Control] = {
+      persistEvent: (IssuedForDeliveryLatest) => Unit,
+      sendCommProgressedEvent: (IssuedForDeliveryLatest => Future[_])
+  )(implicit actorSystem: ActorSystem,
+    ec: ExecutionContext,
+    upcast: T => IssuedForDeliveryLatest): RunnableGraph[Consumer.Control] = {
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/event/gateway/manager/kafka/IssuedForDeliveryGraph.scala]:
```
Welcome to the Ammonite Repl 0.8.2
(Scala 2.12.1 Java 1.8.0_121)
@ implicitly[String => String]
res0: String => String = <function1>
@ res0("abc")
res1: String = "abc"
```

Wow, I didn't know `identity` was available implicitly like that. I wonder where that is coming from. I can't find it in Predef.

File: src/main/scala/com/ovoenergy/event/gateway/manager/kafka/Producer.scala [hunk 7]
```diff
@@ -3,33 +3,36 @@ package com.ovoenergy.event.gateway.manager.kafka
...
 import scala.concurrent.{ExecutionContext, Future}
 
 object Producer extends LoggingWithMDC {
 
-  def send[T <: HasMetadata: SchemaFor: ToRecord](config: Config, topic: String)(
+  def send[T: SchemaFor: ToRecord](config: Config, topic: String)(
       implicit ec: ExecutionContext,
-      scheduler: Scheduler): T => Future[RecordMetadata] = {
+      scheduler: Scheduler,
+      metadataLens: Lens[T, MetadataV2]): T => Future[RecordMetadata] = {
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/event/gateway/manager/kafka/Producer.scala]: Good idea using a lens here 👍

File: src/main/scala/com/ovoenergy/event/gateway/manager/kafka/Producer.scala [hunk 8]
```diff
@@ -3,33 +3,36 @@ package com.ovoenergy.event.gateway.manager.kafka
...
     val producer = KafkaProducer(Conf(new StringSerializer, avroSerializer[T], config.hosts))
     (event: T) =>
       {
+        val metadata = metadataLens.get(event)
         Retry.retryAsync(
           config = config.retry,
-          onFailure = e =>
-            logWarn(Map(TraceTokenMDC -> event.metadata.traceToken), s"Failed to send Kafka event to topic $topic", e)
+          onFailure =
+            e => logWarn(Map(TraceTokenMDC -> metadata.traceToken), s"Failed to send Kafka event to topic $topic", e)
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/event/gateway/manager/kafka/Producer.scala]: If you add the bound `T <: LoggableEvent`, you can include some more useful stuff in the MDC

File: src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/DynamoFormats.scala (hunk 2 above)
Comment (tommpy) [inline src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/DynamoFormats.scala]: Agreed, I'll have a look too and see if there's a better way

File: src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/Persistence.scala (hunk 4 above)
Comment (tommpy) [inline src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/Persistence.scala]: Good spot, I don't think it's needed

File: src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/package.scala (hunk 5 above)
Comment (tommpy) [inline src/main/scala/com/ovoenergy/event/gateway/manager/dynamo/package.scala]: How about a coproduct?

File: src/main/scala/com/ovoenergy/event/gateway/manager/kafka/IssuedForDeliveryGraph.scala (hunk 6 above)
Comment (tommpy) [inline src/main/scala/com/ovoenergy/event/gateway/manager/kafka/IssuedForDeliveryGraph.scala]: Looks like it used to be in predef a few versions ago, then it mysteriously vanished.

File: src/main/scala/com/ovoenergy/event/gateway/manager/kafka/Producer.scala (hunk 8 above)
Comment (tommpy) [inline src/main/scala/com/ovoenergy/event/gateway/manager/kafka/Producer.scala]: Will do

File: src/main/scala/com/ovoenergy/event/gateway/manager/kafka/IssuedForDeliveryGraph.scala (hunk 6 above)
Comment (cb372) [inline src/main/scala/com/ovoenergy/event/gateway/manager/kafka/IssuedForDeliveryGraph.scala]: Update: I got some useful responses to this https://twitter.com/cbirchall/status/862350125670289409

File: src/main/scala/com/ovoenergy/event/gateway/manager/Main.scala [hunk 9]
```diff
@@ -23,9 +22,12 @@ import com.ovoenergy.event.gateway.manager.scheduling.Scheduler
 import com.ovoenergy.event.gateway.manager.twilio.StatusUpdateProcess
 import com.typesafe.config.ConfigFactory
 import org.slf4j.LoggerFactory
-import com.ovoenergy.comms.serialisation.Codecs._
+import com.ovoenergy.comms.serialisation.Serialisation._
 import com.ovoenergy.event.gateway.manager.event._
 import com.ovoenergy.event.gateway.manager.event.EventConversion
+import shapeless.ops.product.ToRecord
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/event/gateway/manager/Main.scala]: looks like this weird import slipped in accidentally

File: src/main/scala/com/ovoenergy/event/gateway/manager/Main.scala [hunk 10]
```diff
@@ -109,7 +111,7 @@ object Main extends App with LoggingWithMDC {
       case _ =>
         logInfo(
           Map(TraceTokenMDC -> issuedForDelivery.metadata.traceToken,
-              ChannelMDC    -> issuedForDelivery.channel.toString,
+              ChannelMDC    -> issuedForDelivery.toString,
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/event/gateway/manager/Main.scala]: why this change? doesn't look right to me

File: src/test/scala/com/ovoenergy/event/gateway/manager/sqs/TwilioStatusUpdatesSpec.scala [hunk 11]
```diff
@@ -50,7 +50,7 @@ class TwilioStatusUpdatesSpec
     while (times < 5 && updatesProcessed.size != 2) {
       TwilioStatusUpdates.receiveAndProcessEvents(config, collectReads)(global.prepare)().futureValue
       times += 1
-      Thread.sleep(300)
+      Thread.sleep(500)
```
Comment (cb372) [inline src/test/scala/com/ovoenergy/event/gateway/manager/sqs/TwilioStatusUpdatesSpec.scala]: beautiful!

File: src/main/scala/com/ovoenergy/event/gateway/manager/Main.scala [hunk 12]
```diff
@@ -98,6 +99,7 @@ object Main extends App with LoggingWithMDC {
   )
   val gatewayIdentifierPersistence = new Persistence(Persistence.Context(dynamoClient, dynamoTable, dynamoRetryConfig))
 
+  implicit val foo             = ToRecord[EmailProgressedLatest]
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/event/gateway/manager/Main.scala]: This foo isn't doing anything. It's a shapeless `ToRecord`, which we don't use anywhere. We use avro4s `ToRecord`, but you shouldn't need to define it yourself.

Comment (tommpy) [inline src/main/scala/com/ovoenergy/event/gateway/manager/Main.scala]: Ooops, that slipped in when I was trying to figure out the implicits

File: src/main/scala/com/ovoenergy/event/gateway/manager/Main.scala (hunk 10 above)
Comment (tommpy) [inline src/main/scala/com/ovoenergy/event/gateway/manager/Main.scala]: Good spot

**Comment by ovo-comms-circleci** [comment]
This PR was successfully deployed to production by [CircleCI build 153](https://circleci.com/gh/ovotech/comms-gateway-event-manager/153) at 2017-05-15 15:51 BST.
